### PR TITLE
octopus: librbd: track complete async operation requests

### PR DESCRIPTION
**Pull request by trociny**
backport tracker: https://tracker.ceph.com/issues/47427

---

backport of https://github.com/ceph/ceph/pull/37012
parent tracker: https://tracker.ceph.com/issues/46803

this backport was staged using ceph-backport.sh version 16.0.0.6848
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Review by idryomov — COMMENTED** [review]
Should https://github.com/ceph/ceph/pull/37316 be brought in as well or is it only relevant for quiescing which isn't in octopus?

**Comment by trociny** [comment]
> Should #37316 be brought in as well or is it only relevant for quiescing which isn't in octopus?

It indeed aimed to fix the issue we observed only for quiesce tests, and it is probably not important for octopus, but I think it will not hurt to backport the relevant parts here too. Added.

**Comment by trociny** [comment]
> Added.

It does not build. I will need to improve.

**Comment by trociny** [comment]
I had to tweak the last patch to use `!=` `AsyncRequestId` operator instead of `==`. In both master and octopus the `==` operator is not explicitly defined, only `!<` and `<`. It seems enough in newer releases but the build fails in octopus.

Alternatively I could explicitly define `operator==` for `AsyncRequestId` in WatchNotifyTypes.h.

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
jenkins test docs

File: src/librbd/ImageWatcher.cc [hunk 1]
```diff
@@ -489,6 +489,46 @@ void ImageWatcher<I>::notify_lock_owner(const Payload& payload,
...
+
+  if (!m_async_complete.insert({id, r}).second) {
+    for (it = m_async_complete_expiration.begin();
+         it != m_async_complete_expiration.end(); it++) {
+      if (it->second != id) {
+        continue;
+      } else {
+        m_async_complete_expiration.erase(it);
+        break;
+      }
```
Comment (idryomov) [inline src/librbd/ImageWatcher.cc:522]:
```suggestion
      if (!(it->second != id)) {
        m_async_complete_expiration.erase(it);
        break;
      }
```

**Comment by idryomov** [comment]
> I had to tweak the last patch to use `!=` `AsyncRequestId` operator instead of `==`. In both master and octopus the `==` operator is not explicitly defined, only `!<` and `<`. It seems enough in newer releases but the build fails in octopus.

Yeah, they are default-defined in C++20 and it could be that newer compilers just do it unconditionally -- I don't think we compile with `-std=c++20`, not even the master branch.

> 
> Alternatively I could explicitly define `operator==` for `AsyncRequestId` in WatchNotifyTypes.h.

It is a good practice to define both if one is defined.  However given that `==` isn't in master, perhaps just open code it like I suggested (just a one line fixup instead of `if`/`else` and `continue`)?

File: src/librbd/ImageWatcher.cc (hunk 1 above)
Comment (trociny) [reply src/librbd/ImageWatcher.cc:522]: Actually it was my version but it looked unreadable and might make wonder an occasional reader.  But ok. I will use it.

**Review by idryomov — APPROVED** [review]

**Pull request merged**